samcli/commands/_utils/custom_options/hook_package_id_option.py

```diff
@@ -28,73 +33,91 @@ def __init__(self, *args, **kwargs):
 
     def handle_parse_result(self, ctx, opts, args):
         opt_name = self.name.replace("_", "-")
-        if self.name in opts:
-            command_name = ctx.command.name
-            if command_name in ["invoke", "start-lambda", "start-api"]:
-                command_name = f"local {command_name}"
-            # validate the hook_package_id value exists
-            hook_package_id = opts[self.name]
-            iac_hook_wrapper = None
-            try:
-                iac_hook_wrapper = IacHookWrapper(hook_package_id)
-            except InvalidHookWrapperException as e:
+        if self.name not in opts:
+            return super().handle_parse_result(ctx, opts, args)
+        command_name = ctx.command.name
+        if command_name in ["invoke", "start-lambda", "start-api"]:
+            command_name = f"local {command_name}"
+        # validate the hook_package_id value exists
+        hook_package_id = opts[self.name]
+        iac_hook_wrapper = None
+        try:
+            iac_hook_wrapper = IacHookWrapper(hook_package_id)
+        except InvalidHookWrapperException as e:
+            raise click.BadParameter(
+                f"{hook_package_id} is not a valid hook package id. This is the list of valid "
+                f"packages ids {get_available_hook_packages_ids()}"
+            ) from e
+
+        self._validate_coexist_options(opt_name, opts)
+
+        _validate_build_command_parameters(command_name, opts)
+
+        if not _check_experimental_flag(hook_package_id, command_name, opts):
+            return super().handle_parse_result(ctx, opts, args)
+
+        self._call_prepare_hook(iac_hook_wrapper, opts)
+        return super().handle_parse_result(ctx, opts, args)
+
+    def _call_prepare_hook(self, iac_hook_wrapper, opts):
+        # call prepare hook
+        built_template_path = DEFAULT_BUILT_TEMPLATE_PATH
+        if not self._force_prepare and os.path.exists(built_template_path):
+            LOG.info("Skip Running Prepare hook. The current application is already prepared.")
```

Comment:
Skipped prepare hook. Current application is already prepared.

Reply:
done